File: riveredge-backend/src/apps/kuaizhizao/schemas/purchase_inquiry.py
"""
采购询价单 Schema

Author: RiverEdge Team
Date: 2026-05-28
"""

from datetime import date, datetime
from decimal import Decimal
from typing import Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator


class PurchaseInquiryItemBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    material_id: int
    material_code: str
    material_name: str
    material_spec: Optional[str] = None
    unit: str = "件"
    quantity: Decimal = Field(..., gt=0)
    required_date: Optional[date] = None
    source_requisition_item_id: Optional[int] = None
    notes: Optional[str] = None


class PurchaseInquiryItemCreate(PurchaseInquiryItemBase):
    pass


class PurchaseInquiryItemResponse(PurchaseInquiryItemBase):
    id: int
    inquiry_id: int
    tenant_id: int
    awarded_supplier_id: Optional[int] = None
    awarded_quote_item_id: Optional[int] = None
    purchase_order_id: Optional[int] = None
    purchase_order_item_id: Optional[int] = None
    created_at: datetime
    updated_at: datetime


class PurchaseInquiryVendorBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    supplier_id: int
    supplier_name: str
    notes: Optional[str] = None


class PurchaseInquiryVendorCreate(PurchaseInquiryVendorBase):
    pass


class PurchaseInquiryVendorResponse(PurchaseInquiryVendorBase):
    id: int
    inquiry_id: int
    tenant_id: int
    status: str
    quoted_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class PurchaseSupplierQuoteItemBase(BaseModel):
    inquiry_item_id: int
    quoted_quantity: Decimal = Field(default=Decimal(0), ge=0)
    unit_price: Decimal = Field(default=Decimal(0), ge=0)
    delivery_date: Optional[date] = None
    lead_time_days: Optional[int] = None
    notes: Optional[str] = None


class PurchaseSupplierQuoteItemResponse(PurchaseSupplierQuoteItemBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    quote_id: int
    tenant_id: int
    is_awarded: bool = False
    created_at: datetime
    updated_at: datetime


class PurchaseSupplierQuoteResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    tenant_id: int
    inquiry_id: int
    supplier_id: int
    supplier_name: str
    quote_code: Optional[str] = None
    quote_date: Optional[date] = None
    valid_until: Optional[date] = None
    status: str
    submission_channel: str
    entered_by: Optional[int] = None
    total_amount: Decimal
    notes: Optional[str] = None
    items: List[PurchaseSupplierQuoteItemResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class UpsertSupplierQuoteRequest(BaseModel):
    supplier_id: int
    supplier_name: Optional[str] = None
    quote_date: Optional[date] = None
    valid_until: Optional[date] = None
    notes: Optional[str] = None
    items: List[PurchaseSupplierQuoteItemBase] = Field(default_factory=list)


class PurchaseInquiryCreate(BaseModel):
    inquiry_code: Optional[str] = None
    inquiry_name: Optional[str] = None
    inquiry_date: Optional[date] = None
    quote_deadline: Optional[date] = None
    source_type: Optional[str] = None
    source_id: Optional[int] = None
    source_code: Optional[str] = None
    notes: Optional[str] = None
    attachments: Optional[list] = None
    items: List[PurchaseInquiryItemCreate] = Field(default_factory=list)
    vendors: List[PurchaseInquiryVendorCreate] = Field(default_factory=list)


class PurchaseInquiryUpdate(BaseModel):
    inquiry_name: Optional[str] = None
    inquiry_date: Optional[date] = None
    quote_deadline: Optional[date] = None
    notes: Optional[str] = None
    attachments: Optional[list] = None
    items: Optional[List[PurchaseInquiryItemCreate]] = None
    vendors: Optional[List[PurchaseInquiryVendorCreate]] = None


class PurchaseInquiryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    tenant_id: int
    inquiry_code: str
    inquiry_name: Optional[str] = None
    inquiry_date: Optional[date] = None
    quote_deadline: Optional[date] = None
    status: str
    buyer_id: Optional[int] = None
    buyer_name: Optional[str] = None
    source_type: Optional[str] = None
    source_id: Optional[int] = None
    source_code: Optional[str] = None
    reviewer_id: Optional[int] = None
    reviewer_name: Optional[str] = None
    review_time: Optional[datetime] = None
    review_status: str
    review_remarks: Optional[str] = None
    total_quantity: Decimal
    notes: Optional[str] = None
    attachments: Optional[list] = None
    created_at: datetime
    updated_at: datetime
    items: List[PurchaseInquiryItemResponse] = Field(default_factory=list)
    vendors: List[PurchaseInquiryVendorResponse] = Field(default_factory=list)
    quotes: List[PurchaseSupplierQuoteResponse] = Field(default_factory=list)
    lifecycle: Optional[dict] = None


class CreateFromRequisitionRequest(BaseModel):
    item_ids: List[int] = Field(..., description="采购申请行ID")
    supplier_ids: Optional[List[int]] = Field(None, description="受邀供应商，不传则按物料默认供应商推荐")
    inquiry_name: Optional[str] = None
    quote_deadline: Optional[date] = None
    notes: Optional[str] = None


class AwardQuotesRequest(BaseModel):
    awards: List[Dict[str, int]] = Field(
        ...,
        description="定标列表，每项含 inquiry_item_id 与 quote_item_id",
    )

    @field_validator("awards", mode="before")
    @classmethod
    def _normalize_awards(cls, v):
        if not v:
            return v
        out = []
        for row in v:
            if not isinstance(row, dict):
                continue
            out.append({
                "inquiry_item_id": int(row["inquiry_item_id"]),
                "quote_item_id": int(row["quote_item_id"]),
            })
        return out


class ConvertInquiryToPORequest(BaseModel):
    item_ids: Optional[List[int]] = Field(None, description="询价行ID，不传则转全部已定标行")
    persist_default_supplier_to_material: bool = False


class ApproveInquiryRequest(BaseModel):
    approved: bool = True
    review_remarks: Optional[str] = None


class ComparisonCell(BaseModel):
    quote_item_id: Optional[int] = None
    quote_id: Optional[int] = None
    supplier_id: Optional[int] = None
    supplier_name: Optional[str] = None
    unit_price: Optional[Decimal] = None
    quoted_quantity: Optional[Decimal] = None
    delivery_date: Optional[date] = None
    lead_time_days: Optional[int] = None
    is_lowest_price: bool = False
    is_awarded: bool = False


class ComparisonRow(BaseModel):
    inquiry_item_id: int
    material_id: int
    material_code: str
    material_name: str
    quantity: Decimal
    required_date: Optional[date] = None
    cells: List[ComparisonCell] = Field(default_factory=list)


class ComparisonMatrixResponse(BaseModel):
    inquiry_id: int
    suppliers: List[PurchaseInquiryVendorResponse] = Field(default_factory=list)
    rows: List[ComparisonRow] = Field(default_factory=list)

File: riveredge-frontend/src/apps/kuaizhizao/services/purchase-inquiry.ts
/**
 * 采购询价单 API
 */

import { apiRequest } from '../../../services/api';
import type { DocumentPushPreview } from './purchase-requisition';

export interface ActionCapability {
  allowed: boolean;
  reason?: string | null;
}

export interface PurchaseInquiryCapabilities {
  update?: ActionCapability;
  delete?: ActionCapability;
  submit?: ActionCapability;
  withdraw_submit?: ActionCapability;
  approve?: ActionCapability;
  revoke_approval?: ActionCapability;
  push_purchase_order?: ActionCapability;
}

export interface PurchaseInquiryItem {
  id?: number;
  inquiry_id?: number;
  material_id?: number;
  material_code?: string;
  material_name?: string;
  material_spec?: string;
  unit?: string;
  quantity?: number;
  required_date?: string;
  source_requisition_item_id?: number;
  awarded_supplier_id?: number;
  awarded_quote_item_id?: number;
  purchase_order_id?: number;
  notes?: string;
}

export interface PurchaseInquiryVendor {
  id?: number;
  inquiry_id?: number;
  supplier_id?: number;
  supplier_name?: string;
  status?: string;
  quoted_at?: string;
}

export interface PurchaseSupplierQuoteItem {
  id?: number;
  quote_id?: number;
  inquiry_item_id?: number;
  quoted_quantity?: number;
  unit_price?: number;
  delivery_date?: string;
  lead_time_days?: number;
  is_awarded?: boolean;
  notes?: string;
}

export interface PurchaseSupplierQuote {
  id?: number;
  inquiry_id?: number;
  supplier_id?: number;
  supplier_name?: string;
  quote_date?: string;
  valid_until?: string;
  status?: string;
  submission_channel?: string;
  total_amount?: number;
  items?: PurchaseSupplierQuoteItem[];
}

export interface PurchaseInquiry {
  id?: number;
  inquiry_code?: string;
  inquiry_name?: string;
  inquiry_date?: string;
  quote_deadline?: string;
  status?: string;
  buyer_id?: number;
  buyer_name?: string;
  source_type?: string;
  source_id?: number;
  source_code?: string;
  review_status?: string;
  items_count?: number;
  total_quantity?: number;
  total_amount?: number;
  notes?: string;
  items?: PurchaseInquiryItem[];
  vendors?: PurchaseInquiryVendor[];
  quotes?: PurchaseSupplierQuote[];
  lifecycle?: Record<string, unknown>;
  capabilities?: PurchaseInquiryCapabilities;
}

export interface ComparisonCell {
  quote_item_id?: number;
  quote_id?: number;
  supplier_id?: number;
  supplier_name?: string;
  unit_price?: number;
  quoted_quantity?: number;
  delivery_date?: string;
  lead_time_days?: number;
  is_lowest_price?: boolean;
  is_awarded?: boolean;
}

export interface ComparisonRow {
  inquiry_item_id: number;
  material_id: number;
  material_code: string;
  material_name: string;
  quantity: number;
  required_date?: string;
  cells: ComparisonCell[];
}

export async function listPurchaseInquiries(params: {
  skip?: number;
  limit?: number;
  lifecycle_stage?: string;
  keyword?: string;
  inquiry_code?: string;
  inquiry_name?: string;
  source_code?: string;
  quote_deadline_from?: string;
  quote_deadline_to?: string;
  created_start_date?: string;
  created_end_date?: string;
  source_id?: number;
  order_by?: string;
  /** 明细表格视图才需要附带 items；普通列表勿开 */
  include_items?: boolean;
} = {}): Promise<{ data: PurchaseInquiry[]; total: number; success: boolean }> {
  return apiRequest('/apps/kuaizhizao/purchase-inquiries', { method: 'GET', params });
}

export async function getPurchaseInquiry(id: number): Promise<PurchaseInquiry> {
  return apiRequest(`/apps/kuaizhizao/purchase-inquiries/${id}`);
}

export async function createPurchaseInquiry(data: Partial<PurchaseInquiry>): Promise<PurchaseInquiry> {
  return apiRequest('/apps/kuaizhizao/purchase-inquiries', { method: 'POST', data });
}

export async function updatePurchaseInquiry(id: number, data: Partial<PurchaseInquiry>): Promise<PurchaseInquiry> {
  return apiRequest(`/apps/kuaizhizao/purchase-inquiries/${id}`, { method: 'PUT', data });
}

export async function deletePurchaseInquiry(id: number): Promise<void> {
  return apiRequest(`/apps/kuaizhizao/purchase-inquiries/${id}`, { method: 'DELETE' });
}

export async function createInquiryFromRequisition(
  requisitionId: number,
  data: { item_ids: number[]; supplier_ids?: number[]; inquiry_name?: string; quote_deadline?: string; notes?: string },
): Promise<PurchaseInquiry> {
  return apiRequest(`/apps/kuaizhizao/purchase-requisitions/${requisitionId}/push-to-purchase-inquiry`, { method: 'POST', data });
}

export type PurchaseInquiryPurchasePullLine = {
  id: number;
  inquiry_id: number;
  inquiry_code?: string;
  material_id?: number;
  material_code?: string;
  material_name?: string;
  material_spec?: string | null;
  unit?: string;
  suggested_quantity?: number;
  pushed_quantity?: number;
  remaining_quantity?: number;
  required_date?: string | null;
  awarded_supplier_id?: number | null;
};

export async function listPurchaseInquiryPurchaseOrderPullLines(params: {
  skip?: number;
  limit?: number;
  keyword?: string;
  inquiry_id?: number;
  pullable_only?: boolean;
}): Promise<{ data: PurchaseInquiryPurchasePullLine[]; total: number }> {
  return apiRequest('/apps/kuaizhizao/purchase-inquiries/purchase-order-pull-lines', {
    method: 'GET',
    params,
  });
}

export async function pullPurchaseInquiryFromRequisitionItems(
  selectedItemIds: number[],
): Promise<PurchaseInquiry> {
  return apiRequest('/apps/kuaizhizao/purchase-inquiries/pull-from-requisition-items', {
    method: 'POST',
    data: { selected_item_ids: selectedItemIds },
  });
}

export async function pullPurchaseOrdersFromInquiryItems(
  selectedItemIds: number[],
): Promise<{
  success: boolean;
  message: string;
  purchase_order_id?: number;
  purchase_order_code?: string;
  purchase_orders?: Array<{ purchase_order_id: number; purchase_order_code: string; supplier_id: number }>;
}> {
  return apiRequest('/apps/kuaizhizao/purchase-orders/pull-from-inquiry-items', {
    method: 'POST',
    data: { selected_item_ids: selectedItemIds },
  });
}

export async function publishPurchaseInquiry(id: number): Promise<PurchaseInquiry> {
  return apiRequest(`/apps/kuaizhizao/purchase-inquiries/${id}/publish`, { method: 'POST' });
}

export async function closeInquiryQuoting(id: number): Promise<PurchaseInquiry> {
  return apiRequest(`/apps/kuaizhizao/purchase-inquiries/${id}/close-quoting`, { method: 'POST' });
}

export async function getInquiryComparison(id: number): Promise<{ inquiry_id: number; suppliers: PurchaseInquiryVendor[]; rows: ComparisonRow[] }> {
  return apiRequest(`/apps/kuaizhizao/purchase-inquiries/${id}/comparison`, { method: 'GET' });
}

export async function upsertSupplierQuote(
  inquiryId: number,
  data: {
    supplier_id: number;
    supplier_name?: string;
    quote_date?: string;
    valid_until?: string;
    notes?: string;
    items: Array<{
      inquiry_item_id: number;
      quoted_quantity?: number;
      unit_price?: number;
      delivery_date?: string;
      lead_time_days?: number;
      notes?: string;
    }>;
  },
): Promise<PurchaseSupplierQuote> {
  return apiRequest(`/apps/kuaizhizao/purchase-inquiries/${inquiryId}/supplier-quotes`, { method: 'POST', data });
}

export async function awardInquiryQuotes(
  inquiryId: number,
  awards: Array<{ inquiry_item_id: number; quote_item_id: number }>,
): Promise<PurchaseInquiry> {
  return apiRequest(`/apps/kuaizhizao/purchase-inquiries/${inquiryId}/award`, { method: 'POST', data: { awards } });
}

export async function convertInquiryToPurchaseOrder(
  inquiryId: number,
  data?: { item_ids?: number[]; persist_default_supplier_to_material?: boolean },
): Promise<{ purchase_orders: Array<{ purchase_order_id: number; purchase_order_code: string; supplier_id: number }> }> {
  return apiRequest(`/apps/kuaizhizao/purchase-inquiries/${inquiryId}/convert-to-purchase-order`, { method: 'POST', data: data ?? {} });
}

export async function previewPushInquiryToPurchaseOrder(inquiryId: number): Promise<DocumentPushPreview> {
  return apiRequest(`/apps/kuaizhizao/purchase-inquiries/${inquiryId}/push-to-purchase-order/preview`, {
    method: 'GET',
  });
}

export async function submitPurchaseInquiry(id: number): Promise<PurchaseInquiry> {
  return apiRequest(`/apps/kuaizhizao/purchase-inquiries/${id}/submit`, { method: 'POST' });
}

export async function approvePurchaseInquiry(id: number, approved: boolean, review_remarks?: string): Promise<PurchaseInquiry> {
  return apiRequest(`/apps/kuaizhizao/purchase-inquiries/${id}/approve`, { method: 'POST', data: { approved, review_remarks } });
}

export async function withdrawPurchaseInquiryApproval(id: number): Promise<PurchaseInquiry> {
  return apiRequest(`/apps/kuaizhizao/purchase-inquiries/${id}/withdraw-approval`, { method: 'POST' });
}

/** 撤回提交（待审核 -> 草稿），走统一审核入口 */
export async function withdrawPurchaseInquirySubmit(id: number): Promise<PurchaseInquiry> {
  return apiRequest(`/core/uni-audit/purchase_inquiry/${id}/withdraw`, { method: 'POST' });
}
